giraph-core/src/main/java/org/apache/giraph/utils/ProgressableUtils.java

```diff
@@ -270,8 +270,16 @@ public static void awaitSemaphorePermits(final Semaphore semaphore,
           // Try to get result from the future
           result = entry.getValue().get(
               MSEC_TO_WAIT_ON_EACH_FUTURE, TimeUnit.MILLISECONDS);
-        } catch (InterruptedException | ExecutionException e) {
-          throw new IllegalStateException("Exception occurred", e);
+        } catch (InterruptedException e) {
+          throw new IllegalStateException("Interrupted", e);
+        } catch (ExecutionException e) {
+          // Execution exception wraps the actual cause
+          if (e.getCause() instanceof RuntimeException) {
```

Comment:
Apparently a null check is not needed, http://stackoverflow.com/questions/2950319/is-null-check-needed-before-calling-instanceof